src/gsym/parser.rs

```diff
@@ -0,0 +1,330 @@
+
+/// Hold the major parts of a standalone GSYM file.
+///
+/// GsymContext provides functions to access major entities in GSYM.
+/// GsymContext can find respective AddressInfo for an address.  But,
+/// it doesn't parse AddressData to get line numbers.
+pub struct GsymContext<'a> {
+    header: Header,
+    addr_tab: &'a [u8],
+    addr_data_off_tab: &'a [u8],
+    str_tab: &'a [u8],
+    raw_data: &'a [u8],
+}
+
+impl<'a> GsymContext<'a> {
+    /// Parse the Header of a standalone GSYM file.
+    ///
+    /// # Arguments
+    ///
+    /// * `data` - is the content of a standalone GSYM.
+    ///
+    /// Returns a GsymContext, which includes the Header and other important tables.
+    pub fn parse_header<'d>(data: &'d [u8]) -> Result<GsymContext<'d>, Error> {
+        let mut off = 0;
+        // Parse Header
+        let magic = decode_uword(data);
+        if magic != GSYM_MAGIC {
+            return Err(Error::new(ErrorKind::InvalidData, "invalid magic number"));
+        }
+        off += 4;
+        let version = decode_uhalf(&data[off..]);
+        if version != GSYM_VERSION {
+            return Err(Error::new(ErrorKind::InvalidData, "unknown version number"));
+        }
+        off += 2;
+        let addr_off_size = data[off];
+        off += 1;
+        let uuid_size = data[off];
+        off += 1;
+        let base_address = decode_udword(&data[off..]);
+        off += 8;
+        let num_addrs = decode_uword(&data[off..]);
+        off += 4;
+        let strtab_offset = decode_uword(&data[off..]);
+        off += 4;
+        let strtab_size = decode_uword(&data[off..]);
+        off += 4;
+        let uuid: [u8; 20] = (&data[off..(off + 20)])
+            .try_into()
+            .expect("input data is too short");
+        off += 20;
+
+        // Get the slices of the Address Table, Address Data Offset Table,
+        // and String table.
+        let addr_tab = &data[off..(off + num_addrs as usize * addr_off_size as usize)];
+        off += num_addrs as usize * addr_off_size as usize;
+        let addr_data_off_tab = &data[off..(off + num_addrs as usize * 4)];
+        off += num_addrs as usize * 4;
+        let str_tab =
+            &data[strtab_offset as usize..(strtab_offset as usize + strtab_size as usize)];
+
+        Ok(GsymContext {
+            header: Header {
+                magic,
+                version,
+                addr_off_size,
+                uuid_size,
+                base_address,
+                num_addrs,
+                strtab_offset,
+                strtab_size,
+                uuid,
+            },
+            addr_tab,
+            addr_data_off_tab,
+            str_tab,
+            raw_data: data,
+        })
+    }
+
+    #[inline(always)]
+    pub fn num_addresses(&self) -> usize {
+        self.header.num_addrs as usize
+    }
+
+    /// Get the address of the an entry in the Address Table.
+    ///
+    /// # Saftety
+    ///
+    /// The code will crash with an invalid index.
```

Comment:
That's not what Rust typically refers to as safety. Panic behavior is usually described in a `Panics` section.